### PR TITLE
Added JSON type to MySQL and PostgreSQL schema

### PR DESCRIPTION
**Pull request by undefinedor**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13019 #10436

### CODE REVIEW

**Comment by Renkas** [comment]
JSON data type is available from MySQL 5.7.8. What happens if I use 5.6?

**Comment by undefinedor** [comment]
@Renkas  throw a exception . Mysql 5.6 does't support JSON.

**Comment by undefinedor** [comment]
@Renkas  throw a exception . Mysql 5.6 does't support JSON.

**Comment by RSalo** [comment]
@Renkas i think Yii should support that, but checking availability this functionality fall on users shoulders

**Comment by samdark** [comment]
Throwing exception sounds OK to me in this case. MySQL 5.6 is 2013.

**Review by dynasource — CHANGES_REQUESTED** [review]
shouldnt the added support for JSON be a little bigger than what this PR contains?

**Comment by samdark** [comment]
@dynasource what's missing?

**Comment by sergeymakinen** [comment]
Even basic JSON support is good but it'd be good to reflect in the title that the PR adds a JSON support to the column schema builder to not get hopes up.

**Comment by dynasource** [comment]
> Even basic JSON support is good but it'd be good to reflect in the title that the PR adds a JSON support to the column schema builder to not get hopes up.

exactly

File: framework/db/pgsql/Schema.php [hunk 1]
```diff
@@ -108,7 +108,7 @@ class Schema extends \yii\db\Schema
         'unknown' => self::TYPE_STRING,
 
         'uuid' => self::TYPE_STRING,
-        'json' => self::TYPE_STRING,
+        'json' => self::TYPE_JSON,
         'jsonb' => self::TYPE_STRING,
```
Comment (cebe) [inline framework/db/pgsql/Schema.php]: same should apply for jsonb, the difference is in the storage of the data, not in the data format. https://www.postgresql.org/docs/9.4/static/datatype-json.html

**Review by cebe — CHANGES_REQUESTED** [review]
needs a changelog, and ideally a test also.

**Comment by jeesus** [comment]
Any estimate when we could see this in stable release?

**Comment by samdark** [comment]
No. Without tests it won't make it to stable release.

File: framework/db/mysql/QueryBuilder.php [hunk 2]
```diff
@@ -43,6 +43,7 @@ class QueryBuilder extends \yii\db\QueryBuilder
         Schema::TYPE_BINARY => 'blob',
         Schema::TYPE_BOOLEAN => 'tinyint(1)',
         Schema::TYPE_MONEY => 'decimal(19,4)',
+        Schema::TYPE_JSON => 'json'
```
Comment (Tigrov) [inline framework/db/mysql/QueryBuilder.php]: It's better to use `Schema::TYPE_JSON => 'jsonb'`
https://www.postgresql.org/docs/current/static/datatype-json.html

File: framework/db/pgsql/QueryBuilder.php [hunk 3]
```diff
@@ -67,6 +67,7 @@ class QueryBuilder extends \yii\db\QueryBuilder
         Schema::TYPE_BINARY => 'bytea',
         Schema::TYPE_BOOLEAN => 'boolean',
         Schema::TYPE_MONEY => 'numeric(19,4)',
+        Schema::TYPE_JSON => 'json',
```
Comment (Tigrov) [inline framework/db/pgsql/QueryBuilder.php]: It's better to use `Schema::TYPE_JSON => 'jsonb'`
https://www.postgresql.org/docs/current/static/datatype-json.html

**Comment by acim** [comment]
What is going on with this?

**Comment by samdark** [comment]
Nothing. With tests missing there's no chance for merging it.

**Comment by nguyenbs** [comment]
Hi,
I see the test is already passed, how about next plan?

**Comment by samdark** [comment]
Existing tests did but there are no tests for the code added.

**Comment by samdark** [comment]
Closed in favor of https://github.com/yiisoft/yii2/pull/15560

**Pull request closed without merging**